test/timestamp.js

```diff
@@ -0,0 +1,220 @@
+  let firestore = new Firestore(
+    Object.assign({}, opts, {
+      projectId: 'test-project',
+      sslCreds: grpc.credentials.createInsecure(),
+    })
+  );
+
+  return firestore._ensureClient().then(() => {
+    firestore._firestoreClient._batchGetDocuments = function() {
+      const stream = through.obj();
+      setImmediate(function() {
+        stream.push({found: document, readTime: {seconds: 5, nanos: 6}});
+        stream.push(null);
+      });
+
+      return stream;
+    };
+
+    return firestore;
+  });
+}
+
+function document(field, value) {
+  let document = {
+    name: `projects/test-project/databases/(default)/documents/coll/doc`,
+    fields: {},
+    createTime: {},
+    updateTime: {},
+  };
+
+  for (let i = 0; i < arguments.length; i += 2) {
+    field = arguments[i];
+    value = arguments[i + 1];
+    document.fields[field] = value;
+  }
+
+  return document;
+}
+
+const DOCUMENT_WITH_TIMESTAMP = document('moonLanding', {
+  valueType: 'timestampValue',
+  timestampValue: {
+    nanos: 123000123,
+    seconds: -14182920,
+  },
+});
+
+const DOCUMENT_WITH_EMPTY_TIMESTAMP = document('moonLanding', {
+  valueType: 'timestampValue',
+  timestampValue: {},
+});
+
+describe('timestamps', function() {
+  it('returned when enabled', function() {
+    return createInstance(
+      {timestampsInSnapshots: true},
+      DOCUMENT_WITH_TIMESTAMP
+    ).then(firestore => {
+      const expected = new Firestore.Timestamp(-14182920, 123000123);
+      return firestore
+        .doc('coll/doc')
+        .get()
+        .then(res => {
+          assert.ok(res.data()['moonLanding'].isEqual(expected));
+          assert.ok(res.get('moonLanding').isEqual(expected));
+        });
+    });
+  });
+
+  it('converted to dates when disabled', function() {
+    /* eslint-disable no-console */
+    const oldErrorLog = console.error;
+    let errorMsg = null;
+    // Prevent error message that prompts to enable `timestampsInSnapshots`
+    // behavior.
+    console.error = msg => {
+      errorMsg = msg;
+    };
+
+    return createInstance(
+      {timestampsInSnapshots: false},
+      DOCUMENT_WITH_TIMESTAMP
+    ).then(firestore => {
+      return firestore
+        .doc('coll/doc')
+        .get()
+        .then(res => {
+          assert.ok(is.date(res.data()['moonLanding']));
+          assert.ok(is.date(res.get('moonLanding')));
+          assert.ok(errorMsg.indexOf('timestampsInSnapshots') !== -1);
+          console.error = oldErrorLog;
+        });
+    });
+    /* eslint-enable no-console */
+  });
+
+  it('retain seconds and nanoseconds', function() {
+    return createInstance(
+      {timestampsInSnapshots: true},
+      DOCUMENT_WITH_TIMESTAMP
+    ).then(firestore => {
+      return firestore
+        .doc('coll/doc')
+        .get()
+        .then(res => {
+          const timestamp = res.get('moonLanding');
+          assert.equal(timestamp.seconds, -14182920);
+          assert.equal(timestamp.nanoseconds, 123000123);
```

Comment:
In client SDKs, I have deliberately added code to truncate a timestamp to microseconds as soon as it is given to Firestore, so that truncation works the same offline and online. Is there a reason not to do the same in server SDKs?

Reply:
Ah, I forgot that server SDKs don't have a local mode. Sorry about that.